src/blog/0.3/linear-types-borrowing.vmd

```diff
@@ -0,0 +1,673 @@
+
+
+<<<<
+It goes from this:
+
+```
+r = RandInt(rand);
+```
+
+where `rand` is a `&Random` reference...
+////
+...to this:
+
+```
+[r, set rand] = RandInt(rand);
+```
+
+where `rand` is an actual `Random`, not just a reference.
+>>>>
+
+
+Now, that line:
+
+ * Moves the `rand` instance into the `RandInt` call.
+ * Uses the square braces to break apart the returned tuple into its two elements:
+    * The `i64` part assigned to the local variable `r`.
+    * The `Random` part is put into the `rand` variable via `set rand`.
+
+
+Here's the rest of `MakeBoard`, for context.
+
+
+<<<<
+It went from this:
+
+```
+// Returns a 2D array of random bools.
+func MakeBoard(
+    rand &Random,
+    num_rows int,
+    num_cols int)
+[][]bool {
+  rows = [][]bool(num_rows);
+  foreach row_i in 0..num_rows {
+    row = []bool(num_cols);
+    foreach col_i in 0..num_cols {
+      r = RandInt(rand);
+      row.push((r mod 2) == 0);
+    }
+    rows.push(row);
+  }
+  return rows;
+}
+```
+////
+...to this:
+
+```
+// Returns a 2D array of random bools.
+func MakeBoard(
+   rand Random,
+   num_rows int,
+   num_cols int)
+([][]bool, Random) {
+  rows = [][]bool(num_rows);
+  foreach row_i in 0..num_rows {
+    row = []bool(num_cols);
+    foreach col_i in 0..num_cols {
+      [r, set rand] = RandInt(rand);
+      row.push((r mod 2) == 0);
+    }
+    rows.push(row);
+  }
+  return (rows, rand);
+}
+```
+>>>>
+
+
+Let's break that down:
+
+ * The `rand &Random` parameter became `rand Random`, to take an actual `Random` instance instead of a reference.
+ * The return type `[][]bool` became `([][]bool, Random)`, so we can later return the `Random` instance.
+ * The `[r, set rand] = RandInt(rand);` line changed, as explained above.
+ * `return (rows, rand)` returns the new board and the `Random` instance to the caller.
+
+
+
+Instead of passing a reference, it's like we're "lending" an owned value to a function, which then returns it to us. For now, let's call this maneuver *own-lending*.
+
+
+If you've used a language like [Rust](https://www.rust-lang.org/) before, this might feel familiar: own-lending is semantically equivalent to borrowing a `&mut Random` and passing it around.
+
+
+Just for fun, I tried applying own-lending to the entire program. Sure enough, the `--print_mem_overhead` flag reported that I reduced generation checks down to *zero.*
+
+
+<slice/>
+
+# Wait, zero?!
+
+This _blew_ my _mind_. That shouldn't be possible. I've never seen `--print_mem_overhead` print zero before!
+
+
+Yet there it was.
+
+
+I sat back from my desk, and got up for a walk. After ten minutes, I realized I'd just been staring at a wall, and hadn't actually walked anywhere.
+
+
+It slowly sank in over the next few days.
+
+
+*Using own-lending, we can completely eliminate every single generation check in a program, to get memory safety without borrow checking, reference counting, or tracing garbage collection.*
```

Comment:
lots of commas, let's try to remove this one.... change ", to get" to "resulting in"